hrms/overrides/expense_claim.py

```diff
@@ -0,0 +1,343 @@
+                if old_doc.status not in [ExpenseClaimConstants.PENDING_APPROVAL_BY_ADMIN]:
+                    frappe.throw(_(f"Invalid State Transition to state {self.state}"))
+                if not user_has_role(frappe.session.user, RoleConstants.OFFICE_ADMIN_ROLE):
+                    frappe.throw(_("Only Admin approve the document"))
+            elif self.status == ExpenseClaimConstants.PENDING_PAYMENT:
+                if old_doc.status not in [ExpenseClaimConstants.PENDING_APPROVAL_BY_HR]:
+                    frappe.throw(_(f"Invalid State Transition to state {self.state}"))
+                if not user_has_role(frappe.session.user, RoleConstants.HR_L2_EXPENSE_ROLE):
+                    frappe.throw(
+                        _("Only user having HR L2 Expense Approver role can approve the document")
+                    )
+            elif self.status == ExpenseClaimConstants.PAID and self.docstatus == 1:
+                if old_doc.status not in [
+                    ExpenseClaimConstants.PENDING_PAYMENT,
+                    ExpenseClaimConstants.PARTLY_CLAIMED,
+                    ExpenseClaimConstants.CLAIMED,
+                ]:
+                    frappe.throw(_(f"Invalid State Transition to state {self.state}"))
+                if old_doc.status in [ExpenseClaimConstants.PENDING_PAYMENT]:
+                    if not user_has_role(frappe.session.user, RoleConstants.FINANCE_ROLE):
+                        frappe.throw(
+                            _(
+                                f"Invalid permission, only user with role {RoleConstants.FINANCE_ROLE} "
+                            )
+                        )
+            elif self.status == ExpenseClaimConstants.CANCELLED:
+                if old_doc.status not in [
+                    ExpenseClaimConstants.DRAFT,
+                    ExpenseClaimConstants.PENDING_PAYMENT,
+                    ExpenseClaimConstants.PENDING_APPROVAL_BY_HR,
+                ]:
+                    frappe.throw(_(f"Invalid State Transition to state {self.state}"))
+                if (
+                    old_doc.status == ExpenseClaimConstants.DRAFT
+                    and not frappe.session.user != self.owner
+                ):
+                    frappe.throw(_(f"Can be Canceled by Owner Only"))
+                elif (
+                    old_doc.status == ExpenseClaimConstants.PENDING_PAYMENT
+                    and frappe.session.user != "Administrator"
+                    and frappe.session.user != self.approver_1
+                ):
+                    frappe.throw(_(f"Can be Canceled either by {self.approver_1} or by Admin"))
+                elif (
+                    old_doc.status == ExpenseClaimConstants.PENDING_APPROVAL_BY_HR
+                    and not user_has_role(frappe.session.user, RoleConstants.HR_L2_EXPENSE_ROLE)
+                ):
+                    frappe.throw(_("Only user having HR L2 Expense Role can cancel the document"))
```

Comment:
add a default else to fail unhandled state transition

Reply:
Done.